// @tanstack/router-generator/dist/cjs/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_config = require("./config.cjs");
const require_rootPathId = require("./filesystem/physical/rootPathId.cjs");
const require_utils = require("./utils.cjs");
const require_getRouteNodes = require("./filesystem/virtual/getRouteNodes.cjs");
const require_getRouteNodes$1 = require("./filesystem/physical/getRouteNodes.cjs");
const require_generator = require("./generator.cjs");
exports.Generator = require_generator.Generator;
exports.baseConfigSchema = require_config.baseConfigSchema;
exports.capitalize = require_utils.capitalize;
exports.checkRouteFullPathUniqueness = require_utils.checkRouteFullPathUniqueness;
exports.cleanPath = require_utils.cleanPath;
exports.configSchema = require_config.configSchema;
exports.determineInitialRoutePath = require_utils.determineInitialRoutePath;
exports.format = require_utils.format;
exports.getConfig = require_config.getConfig;
exports.inferFullPath = require_utils.inferFullPath;
exports.multiSortBy = require_utils.multiSortBy;
exports.physicalGetRouteNodes = require_getRouteNodes$1.getRouteNodes;
exports.removeExt = require_utils.removeExt;
exports.removeLeadingSlash = require_utils.removeLeadingSlash;
exports.removeTrailingSlash = require_utils.removeTrailingSlash;
exports.removeUnderscores = require_utils.removeUnderscores;
exports.replaceBackslash = require_utils.replaceBackslash;
exports.resetRegex = require_utils.resetRegex;
exports.resolveConfigPath = require_config.resolveConfigPath;
exports.rootPathId = require_rootPathId.rootPathId;
exports.routePathToVariable = require_utils.routePathToVariable;
exports.trimPathLeft = require_utils.trimPathLeft;
exports.virtualGetRouteNodes = require_getRouteNodes.getRouteNodes;
exports.writeIfDifferent = require_utils.writeIfDifferent;
